### ISSUE TITLE
x0tigervncserver locks up frequently

### ISSUE THREAD

**Issue by akorn** [TigerVNC/tigervnc#1047]
**Describe the bug**
Intermittently, but about once every few minutes, x0tigervncserver locks up -- the network connection between existing clients and the server stays up, but no input is processed and the view is not refreshed. The server doesn't respond to new connections. `IdleTimeout` doesn't kick in.

**To Reproduce**
Happens randomly, but apparently only during active use, not when the client is idle.

**Client (please complete the following information):**
 - OS: Devuan ceres 64 bit
 - VNC client: tigervnc
 - VNC client version: 1.10.1+dfsg-3 
 - Client downloaded from: shipped by Devuan (likely the same as the the Debian sid package)

**Server (please complete the following information):**
 - OS: Devuan ceres 64bit
 - VNC server: x0tigervncserver
 - VNC server version: 1.10.80-20200605gitda1ce97 (but could also reproduce with 1.10.1+dfsg-3)
 - Server downloaded from: tigervnc.org
 - Server was started using: `ssh otherhost env DISPLAY=:0 x0tigervncserver PasswordFile=/home/myuser/.vnc/passwd -RawKeyboard -IdleTimeout 30 -Log "'*:stderr:100'" -AlwaysShared -AcceptSetDesktopSize=0 -ClientWaitTimeMillis 500`

**Additional context**
Last few lines of server output:

```
Wed Jun 17 16:01:46 2020
 VNCSConnST:  Key released: 0xff54 / 0xd0
 XDesktop:    116 up
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key pressed: 0xff52 / 0xc8
 XDesktop:    111 down
 VNCSConnST:  Key released: 0xff52 / 0xc8
 XDesktop:    111 up
```
Output stops here despite client activity.

stracing the server shows:
```
# strace -ff -s200 -yy -p 12791
strace: Process 12791 attached
restart_syscall(<... resuming interrupted read ...>) = 1
recvfrom(4<UNIX:[7139518->7136768]>, "[\0c-\206\3\0\0\n\0\240\10@s\v\32\0\0\202\5\0\n\36\0\0\0\0\0\200\21\240\5", 9206956, 0, NULL, NULL) = 32
recvfrom(4<UNIX:[7139518->7136768]>, 0xfe5d44, 9206924, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4<UNIX:[7139518->7136768]>, events=POLLIN}], 1, -1) = 1 ([{fd=4, revents=POLLIN}])
recvfrom(4<UNIX:[7139518->7136768]>, "[\0c-\206\3\0\0\n\0\240\10\241]\f\32\0\0\202\5\0\n\36\0\0\0\0\0\200\21\240\5", 9206924, 0, NULL, NULL) = 32
recvfrom(4<UNIX:[7139518->7136768]>, 0xfe5d64, 9206892, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4<UNIX:[7139518->7136768]>, events=POLLIN}], 1, -1) = 1 ([{fd=4, revents=POLLIN}])
recvfrom(4<UNIX:[7139518->7136768]>, "[\0c-\206\3\0\0\n\0\240\10\0H\r\32\0\0\202\5\0\n\36\0\0\0\0\0\200\21\240\5", 9206892, 0, NULL, NULL) = 32
recvfrom(4<UNIX:[7139518->7136768]>, 0xfe5d84, 9206860, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4<UNIX:[7139518->7136768]>, events=POLLIN}], 1, -1) = 1 ([{fd=4, revents=POLLIN}])
recvfrom(4<UNIX:[7139518->7136768]>, "[\0c-\206\3\0\0\n\0\240\10^2\16\32\0\0\202\5\0\n\36\0\0\0\0\0\200\21\240\5", 9206860, 0, NULL, NULL) = 32
recvfrom(4<UNIX:[7139518->7136768]>, 0xfe5da4, 9206828, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4<UNIX:[7139518->7136768]>, events=POLLIN}], 1, -1) = 1 ([{fd=4, revents=POLLIN}])
recvfrom(4<UNIX:[7139518->7136768]>, "[\0c-\206\3\0\0\n\0\240\10\300\34\17\32\0\0\202\5\0\n\36\0\0\0\0\0\200\21\240\5", 9206828, 0, NULL, NULL) = 32
recvfrom(4<UNIX:[7139518->7136768]>, 0xfe5dc4, 9206796, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4<UNIX:[7139518->7136768]>, events=POLLIN}], 1, -1) = 1 ([{fd=4, revents=POLLIN}])
recvfrom(4<UNIX:[7139518->7136768]>, "[\0c-\206\3\0\0\n\0\240\10\36\7\20\32\0\0\202\5\0\n\36\0\0\0\0\0\200\21\240\5", 9206796, 0, NULL, NULL) = 32
recvfrom(4<UNIX:[7139518->7136768]>, 0xfe5de4, 9206764, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4<UNIX:[7139518->7136768]>, events=POLLIN}], 1, -1) = 1 ([{fd=4, revents=POLLIN}])
recvfrom(4<UNIX:[7139518->7136768]>, "[\0c-\206\3\0\0\n\0\240\10\200\361\20\32\0\0\202\5\0\n\36\0\0\0\0\0\200\21\240\5", 9206764, 0, NULL, NULL) = 32
recvfrom(4<UNIX:[7139518->7136768]>, 0xfe5e04, 9206732, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4<UNIX:[7139518->7136768]>, events=POLLIN}], 1, -1) = 1 ([{fd=4, revents=POLLIN}])
recvfrom(4<UNIX:[7139518->7136768]>, "[\0c-\206\3\0\0\n\0\240\10\340\333\21\32\0\0\202\5\0\n\36\0\0\0\0\0\200\21\240\5", 9206732, 0, NULL, NULL) = 32
recvfrom(4<UNIX:[7139518->7136768]>, 0xfe5e24, 9206700, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
poll([{fd=4<UNIX:[7139518->7136768]>, events=POLLIN}], 1, -1
```

**Comment by CendioOssman** [comment]
Thank you for trying to debug this. I'm afraid it's not something we've seen before.

Does the `strace` stop when this happens or does it continue outputting those lines?

And do you think you could attach `gdb` when this happens and get a backtrace for us? You probably also need to install some package with debug symbols. Unfortunately I don't know what those are called in Debian.

**Comment by akorn** [comment]
`strace` continues outputting similar lines indefinitely.

I can try gdb but I'm not very skilled in its use. For example, I don't know how to use the debug symbol packages with it; can you point me at some instructions?

**Comment by CendioOssman** [comment]
> `strace` continues outputting similar lines indefinitely.
> 

Alright. Good to know. I would assume that UNIX socket is to the X server. So it seems it is flooding it with some data. Difficult to say what though, so hopefully gdb can help.

> I can try gdb but I'm not very skilled in its use. For example, I don't know how to use the debug symbol packages with it; can you point me at some instructions?

It should load the debug symbols automatically as long as they are installed.

Once the hang happens, just run `gdb -p <pid of server>`. Then when you get a gdb prompt just type `bt full`.

**Comment by akorn** [comment]
```
(gdb) bt full
#0  0x00007fc77da15b43 in poll () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
#1  0x00007fc77d6ebdd3 in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
No symbol table info available.
#2  0x00007fc77d6ec14f in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
No symbol table info available.
#3  0x00007fc77d6e9f8f in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
No symbol table info available.
#4  0x00007fc77d6ea411 in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
No symbol table info available.
#5  0x00007fc77d6ea4a4 in xcb_writev () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
No symbol table info available.
#6  0x00007fc77e46488e in _XSend () from /usr/lib/x86_64-linux-gnu/libX11.so.6
No symbol table info available.
#7  0x00007fc77e464dbc in _XReply () from /usr/lib/x86_64-linux-gnu/libX11.so.6
No symbol table info available.
#8  0x00007fc77e44a1dc in XGetImage () from /usr/lib/x86_64-linux-gnu/libX11.so.6
No symbol table info available.
#9  0x00007fc77e44a46a in XGetSubImage () from /usr/lib/x86_64-linux-gnu/libX11.so.6
No symbol table info available.
#10 0x0000564d979f8a1c in ShmImage::get (this=<optimized out>, wnd=<optimized out>, x=<optimized out>, y=<optimized out>, w=<optimized out>, h=<optimized out>, dst_x=587, dst_y=1205) at ./unix/x0vncserver/Image.cxx:330
No locals.
#11 0x0000564d979fae3d in XPixelBuffer::grabRect (r=..., this=0x564d98a2f150) at ./common/rfb/Rect.h:108
No locals.
#12 XPixelBuffer::grabRegion (this=0x564d98a2f150, region=...) at ./unix/x0vncserver/XPixelBuffer.cxx:77
        rects = std::vector of length 14, capacity 14 = {{tl = {x = 0, y = 25}, br = {x = 2560, y = 1202}}, {tl = {x = 0, y = 1202}, br = {x = 318, y = 1203}}, {tl = {x = 583, y = 1202}, br = {x = 2560, y = 1203}}, {tl = {x = 0, y = 1203}, br = {x = 316, y = 1204}}, {tl = {x = 585, y = 1203}, br = {x = 2560,
              y = 1204}}, {tl = {x = 0, y = 1204}, br = {x = 315, y = 1205}}, {tl = {x = 586, y = 1204}, br = {x = 2560, y = 1205}}, {tl = {x = 0, y = 1205}, br = {x = 314, y = 1206}}, {tl = {x = 587, y = 1205}, br = {x = 2560, y = 1206}}, {tl = {x = 0, y = 1206}, br = {x = 313, y = 1208}}, {tl = {x = 588,
              y = 1206}, br = {x = 2560, y = 1208}}, {tl = {x = 0, y = 1208}, br = {x = 312, y = 1400}}, {tl = {x = 589, y = 1208}, br = {x = 2560, y = 1400}}, {tl = {x = 0, y = 1400}, br = {x = 2560, y = 1410}}}
        i = {tl = {x = 587, y = 1205}, br = {x = 2560, y = 1206}}
#13 0x0000564d97a0f0c7 in rfb::VNCServerST::writeUpdate (this=0x7ffd1d8ba050) at ./common/rfb/VNCServerST.cxx:796
        ui = {changed = {rgn = 0x564d98c01410}, copied = {rgn = 0x564d98a37390}, copy_delta = {x = 0, y = 0}}
        toCheck = {rgn = 0x564d98c1fc00}
        ci = non-dereferenceable iterator for std::list
        ci_next = non-dereferenceable iterator for std::list
        __PRETTY_FUNCTION__ = "void rfb::VNCServerST::writeUpdate()"
#14 0x0000564d97a0f1ea in rfb::VNCServerST::handleTimeout (this=0x7ffd1d8ba050, t=<optimized out>) at ./common/rfb/VNCServerST.cxx:614
No locals.
#15 0x0000564d97a0c18c in rfb::Timer::checkTimeouts () at ./common/rfb/Timer.cxx:71
        timer = 0x7ffd1d8ba288
        before = {tv_sec = 1594306122, tv_usec = 667682}
        start = {tv_sec = 1594306122, tv_usec = 667682}
#16 0x0000564d979f6b0e in main (argc=<optimized out>, argv=<optimized out>) at ./unix/x0vncserver/x0vncserver.cxx:363
        tv = {tv_sec = 0, tv_usec = 0}
        wfds = {fds_bits = {0 <repeats 16 times>}}
        i = <optimized out>
        wait_ms = <optimized out>
        rfds = {fds_bits = {0 <repeats 16 times>}}
        sockets = std::__cxx11::list = {[0] = 0x564d98a29730}   
        clients_connected = <optimized out>
        n = <optimized out>
        geo = {static m_geometryParam = {<rfb::VoidParameter> = {_vptr.VoidParameter = 0x564d97a5a270 <vtable for rfb::StringParameter+16>, _next = 0x0, immutable = false, name = 0x564d97a3c041 "Geometry",
              description = 0x564d97a3c098 "Screen area shown to VNC clients. Format is <width>x<height>+<offset_x>+<offset_y>, more information in man X, section GEOMETRY SPECIFICATIONS. If the argument is empty, full screen is shown to VNC cl"..., mutex = 0x564d989ef9f0}, value = 0x564d989ef930 "",
            def_value = 0x564d97a3d80f ""}, m_fullWidth = 4480, m_fullHeight = 1440, m_rect = {tl = {x = 0, y = 0}, br = {x = 4480, y = 1440}}}
        hostsData = <optimized out>
        fileTcpFilter = {<network::TcpFilter> = {<network::ConnectionFilter> = {_vptr.ConnectionFilter = 0x564d97a59a68 <vtable for FileTcpFilter+16>}, filter = std::__cxx11::list = {[0] = {action = network::TcpFilter::Reject, address = {u = {sa = {sa_family = 0, sa_data = '\000' <repeats 13 times>}, sin = {
                      sin_family = 0, sin_port = 0, sin_addr = {s_addr = 0}, sin_zero = "\000\000\000\000\000\000\000"}, sin6 = {sin6_family = 0, sin6_port = 0, sin6_flowinfo = 0, sin6_addr = {__in6_u = {__u6_addr8 = '\000' <repeats 15 times>, __u6_addr16 = {0, 0, 0, 0, 0, 0, 0, 0}, __u6_addr32 = {0, 0, 0, 0}}},
                      sin6_scope_id = 0}}}, prefixlen = 0, mask = {u = {sa = {sa_family = 0, sa_data = '\000' <repeats 13 times>}, sin = {sin_family = 0, sin_port = 0, sin_addr = {s_addr = 0}, sin_zero = "\000\000\000\000\000\000\000"}, sin6 = {sin6_family = 0, sin6_port = 0, sin6_flowinfo = 0, sin6_addr = {
                        __in6_u = {__u6_addr8 = '\000' <repeats 15 times>, __u6_addr16 = {0, 0, 0, 0, 0, 0, 0, 0}, __u6_addr32 = {0, 0, 0, 0}}}, sin6_scope_id = 0}}}}}}, fileName = 0x564d989ed7c0 "", lastModTime = 0}
        sched = {m_interval = 30, m_maxload = 35, m_initialState = false, m_passStarted = {m_timeval = {tv_sec = 1594306122, tv_usec = 363512}}, m_ratedDuration = 500, m_sleepStarted = {m_timeval = {tv_sec = 1594306122, tv_usec = 666590}}, m_sleeping = false, m_sleptThisPass = 66, m_errors = {62, 187, 448, 422,
            397, 374, 5, 8}, m_errorSum = 1903, m_errorAbsSum = 1903, m_durations = {92, 217, 478, 452, 427, 404, 35, 38}, m_durationSum = 2143, m_slept = {15, 28, 149, 186, 177, 127, 19, 2}, m_sleptSum = 703, m_idx = 6, m_count = 14}
        desktop = {<rfb::SDesktop> = {<rfb::InputHandler> = {_vptr.InputHandler = 0x564d97a59a90 <vtable for XDesktop+16>}, <No data fields>}, <TXGlobalEventHandler> = {_vptr.TXGlobalEventHandler = 0x564d97a59b20 <vtable for XDesktop+160>}, <QueryResultCallback> = {
            _vptr.QueryResultCallback = 0x564d97a59b38 <vtable for XDesktop+184>}, dpy = 0x564d98a0b2e0, geometry = 0x7ffd1d8b9fe0, pb = 0x564d98a2f150, server = 0x7ffd1d8ba050, queryConnectDialog = 0x0, queryConnectSock = 0x0, oldButtonMask = 0, haveXtest = true, haveDamage = true, maxButtons = 8,
          pressedKeys = std::map with 0 elements, running = true, damage = 69206026, xdamageEventBase = 91, xkbEventBase = 85, xfixesEventBase = 87, xrandrEventBase = 89, outputIdMap = std::map with 2 elements = {[770] = 1804289383, [772] = 846930886}, randrSyncSerial = 0, ledMasks = {4, 2, 1}, ledState = 2,
          codeMap = 0x564d97a3c400 <code_map_qnum_to_xorgevdev>, codeMapLen = 254}
        server = {<rfb::VNCServer> = {<rfb::UpdateTracker> = {_vptr.UpdateTracker = 0x564d97a5a608 <vtable for rfb::VNCServerST+16>}, <network::SocketServer> = {_vptr.SocketServer = 0x564d97a5a6f0 <vtable for rfb::VNCServerST+248>}, <No data fields>}, <rfb::Timer::Callback> = {
            _vptr.Callback = 0x564d97a5a738 <vtable for rfb::VNCServerST+320>}, blacklist = {blm = std::map with 0 elements}, blHosts = 0x7ffd1d8ba068, desktop = 0x7ffd1d8ba470, desktopStarted = true, blockCounter = 0, pb = 0x564d98a2f150, screenLayout = {screens = std::__cxx11::list = {[0] = {id = 1804289383,
                dimensions = {tl = {x = 2560, y = 360}, br = {x = 4480, y = 1440}}, flags = 0}, [1] = {id = 846930886, dimensions = {tl = {x = 0, y = 0}, br = {x = 2560, y = 1440}}, flags = 0}}}, ledState = 2, name = {buf = 0x564d98a2a820 "x0vncserver"}, clients = std::__cxx11::list = {[0] = 0x564d98a36c80},
          pointerClient = 0x0, clipboardClient = 0x0, clipboardRequestors = empty std::__cxx11::list, closingSockets = empty std::__cxx11::list, comparer = 0x564d98bd8800, cursorPos = {x = 818, y = 1439}, cursor = 0x564d98bd8690, renderedCursor = {<rfb::PixelBuffer> = {
              _vptr.PixelBuffer = 0x564d97a5a800 <vtable for rfb::RenderedCursor+16>, format = {bpp = 32, depth = 24, trueColour = true, bigEndian = false, redMax = 255, greenMax = 255, blueMax = 255, redShift = 16, greenShift = 8, blueShift = 0, redBits = 8, greenBits = 8, blueBits = 8, maxBits = 8, minBits = 8,
                endianMismatch = false,
                static upconvTable = "\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377"..., static downconvTable = <same as static member of an already seen type>,
                static _init = {<No data fields>}}, width_ = 4480, height_ = 1440}, buffer = {<rfb::FullFramePixelBuffer> = {<rfb::ModifiablePixelBuffer> = {<rfb::PixelBuffer> = {_vptr.PixelBuffer = 0x564d97a5a520 <vtable for rfb::ManagedPixelBuffer+16>, format = {bpp = 32, depth = 24, trueColour = true,
                      bigEndian = false, redMax = 255, greenMax = 255, blueMax = 255, redShift = 16, greenShift = 8, blueShift = 0, redBits = 8, greenBits = 8, blueBits = 8, maxBits = 8, minBits = 8, endianMismatch = false,
                      static upconvTable = "\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377\000\377"..., static downconvTable = <same as static member of an already seen type>, 
                      static _init = {<No data fields>}}, width_ = 15, height_ = 21}, <No data fields>}, data = 0x564d98bf5330 "\r\r\t", stride = 15}, data_ = 0x564d98bf5330 "\r\r\t", datasize = 1260}, offset = {x = 957, y = 585}}, renderedCursorInvalid = true, 
          keyRemapper = 0x564d97a5e2c0 <rfb::KeyRemapper::defInstance>, idleTimer = {dueTime = {tv_sec = 15481355650793524, tv_usec = 12103604389871656}, timeoutMs = 2949164, cb = 0x7ffd1d8ba060, static pending = {<std::__cxx11::_List_base<rfb::Timer*, std::allocator<rfb::Timer*> >> = {
                _M_impl = {<std::allocator<std::_List_node<rfb::Timer*> >> = {<__gnu_cxx::new_allocator<std::_List_node<rfb::Timer*> >> = {<No data fields>}, <No data fields>}, _M_node = {<std::__detail::_List_node_base> = {_M_next = 0x564d98bf59e0, _M_prev = 0x564d98a2fdd0}, _M_size = 3}}}, <No data fields>}}, 
          disconnectTimer = {dueTime = {tv_sec = 10977687302897700, tv_usec = 7599936041975832}, timeoutMs = 1900572, cb = 0x7ffd1d8ba060, static pending = {<std::__cxx11::_List_base<rfb::Timer*, std::allocator<rfb::Timer*> >> = {
                _M_impl = {<std::allocator<std::_List_node<rfb::Timer*> >> = {<__gnu_cxx::new_allocator<std::_List_node<rfb::Timer*> >> = {<No data fields>}, <No data fields>}, _M_node = {<std::__detail::_List_node_base> = {_M_next = 0x564d98bf59e0, _M_prev = 0x564d98a2fdd0}, _M_size = 3}}}, <No data fields>}}, 
          connectTimer = {dueTime = {tv_sec = 0, tv_usec = -1}, timeoutMs = -256, cb = 0x7ffd1d8ba060, static pending = {<std::__cxx11::_List_base<rfb::Timer*, std::allocator<rfb::Timer*> >> = {
                _M_impl = {<std::allocator<std::_List_node<rfb::Timer*> >> = {<__gnu_cxx::new_allocator<std::_List_node<rfb::Timer*> >> = {<No data fields>}, <No data fields>}, _M_node = {<std::__detail::_List_node_base> = {_M_next = 0x564d98bf59e0, _M_prev = 0x564d98a2fdd0}, _M_size = 3}}}, <No data fields>}}, 
          frameTimer = {dueTime = {tv_sec = 1594306122, tv_usec = 666743}, timeoutMs = 16, cb = 0x7ffd1d8ba060, static pending = {<std::__cxx11::_List_base<rfb::Timer*, std::allocator<rfb::Timer*> >> = {
                _M_impl = {<std::allocator<std::_List_node<rfb::Timer*> >> = {<__gnu_cxx::new_allocator<std::_List_node<rfb::Timer*> >> = {<No data fields>}, <No data fields>}, _M_node = {<std::__detail::_List_node_base> = {_M_next = 0x564d98bf59e0, _M_prev = 0x564d98a2fdd0}, _M_size = 3}}}, <No data fields>}}}
        dpy = 0x564d98a0b2e0
        dpyStr = {buf = 0x564d989ecb80 ""}
        listeners = std::__cxx11::list = {[0] = 0x564d98a2aa60, [1] = 0x564d98a2e640}
```

This is with 38726ce083db1a9227325bf87989513499bfa698.

**Comment by CendioOssman** [comment]
Hm... that trace shows it in the middle of sending an update and seems to be working just fine.

Could this be a client issue? Have you tried more than one client?

**Comment by akorn** [comment]
It can't be purely a client issue because while it's happening I can't connect another client and because the client is not dropped due to `IdleTimeout`.

I haven't yet tried having several clients attached simultaneously to see if it locks up for all of them.

**Comment by CendioOssman** [comment]
Alright, let's try this; Do the `gdb` thing again. But after the `bt` run the command `finish`. It will wait for the current function to exit and then print a line once it is done. Do that as many times as you can and let us know where it stops.

**Comment by balwierz** [comment]
I have a very similar issue using `/usr/bin/vncserver -localhost no -autokill` so I am not starting another issue. (Please amend if you think this is a separate issue)

VNC server display gets randomly frozen. Seems to accept no input, no refreshing.
It happened a couple of times, always during active work after a longer time of `vncserver` running.
Twice, after a longer while (10min+ but not sure exactly), a freeze was gone and I was able to return to work without any discernible problems.
`Xtigervnc` causes 100% CPU utilisation. 
Other users of the same server (all using tigervnc viewer for Mac) did not encounter such problems.

- OS: Debian testing
- tigervnc server version 1.10.1 (dfsg-6 amd64)
- Clients: tigervncviewer (Arch Linux. I checked the two most recent version), realvnc (windows)

Strace shows the following displayed at infinitum:
```
[pid 3522168] --- SIGALRM {si_signo=SIGALRM, si_code=SI_KERNEL} ---
[pid 3522168] rt_sigreturn({mask=[]})   = 8
[pid 3522168] writev(20<UNIX:[327332923->327327076,@"/tmp/.X11-unix/X2"]>, [{iov_base="\34\0\231\376\2\0 \0q\306\10\0\16ZL\326\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", iov_len=32}], 1) = 32
[pid 3522168] writev(6<UNIX:[327254006->327327070,@"/tmp/.X11-unix/X2"]>, [{iov_base="\37\253\vB\0\0\0\0\2\0 \0F\0\0\0J\0\0\0q\306\10\0\20\234\235\3761V\0\0", iov_len=32}], 1) = 32
[pid 3522168] --- SIGALRM {si_signo=SIGALRM, si_code=SI_KERNEL} ---
[pid 3522168] rt_sigreturn({mask=[]})   = 32
```
Socket `/tmp/.X11-unix/X2` exists.

gdb backtrace
```
#0  0x00005631fea424c8 in dixLookupProperty (result=result@entry=0x7fff4f13ab10, pWin=pWin@entry=0x563200b7e4a0, propertyName=propertyName@entry=308152, client=client@entry=0x563200523a40, access_mode=access_mode@entry=2)
    at ../../../../unix/xserver/dix/property.c:99
        pProp = 0x56320716be50
        rc = 8
#1  0x00005631fea42840 in dixChangeWindowProperty (pClient=0x563200523a40, pWin=0x563200b7e4a0, property=property@entry=308152, type=type@entry=74, format=format@entry=8, mode=mode@entry=0, len=58, value=0x56320139a9c0, sendevent=1)
    at ../../../../unix/xserver/dix/property.c:269
        pProp = 0x7fff4f13ab98
        savedProp = <optimized out>
        sizeInBytes = 1
        totalSize = 58
        rc = <optimized out>
        data = <optimized out>
        access_mode = 2
#2  0x00005631fe9d7aa4 in vncConvertSelection (client=0x5632008dfff0, selection=70, target=74, property=308152, requestor=2097154, time=0, data=0x56320139a9c0 "dpkg-query -Wf '${Installed-Size}\\t${Package}\\n' | sort -n")
    at ../../../../../unix/xserver/hw/vnc/vncSelection.c:380
        pSel = 0x563200d56880
        pWin = 0x563200b7e4a0
        rc = 0
        realProperty = 308152
        event = {u = {u = {type = 31 '\037', detail = 171 '\253', sequenceNumber = 16907}, keyButtonPointer = {pad00 = 1108060959, time = 0, root = 2097154, event = 70, child = 74, rootX = -19527, rootY = 4, eventX = -25584, eventY = -355,
              state = 22065, sameScreen = 0 '\000', pad1 = 0 '\000'}, enterLeave = {pad00 = 1108060959, time = 0, root = 2097154, event = 70, child = 74, rootX = -19527, rootY = 4, eventX = -25584, eventY = -355, state = 22065,
              mode = 0 '\000', flags = 0 '\000'}, focus = {pad00 = 1108060959, window = 0, mode = 2 '\002', pad1 = 0 '\000', pad2 = 32 ' ', pad3 = 0 '\000'}, expose = {pad00 = 1108060959, window = 0, x = 2, y = 32, width = 70, height = 0,
              count = 74, pad2 = 0}, graphicsExposure = {pad00 = 1108060959, drawable = 0, x = 2, y = 32, width = 70, height = 0, minorEvent = 74, count = 0, majorEvent = 185 '\271', pad1 = 179 '\263', pad2 = 4 '\004', pad3 = 0 '\000'},
            noExposure = {pad00 = 1108060959, drawable = 0, minorEvent = 2, majorEvent = 32 ' ', bpad = 0 '\000'}, visibility = {pad00 = 1108060959, window = 0, state = 2 '\002', pad1 = 0 '\000', pad2 = 32 ' ', pad3 = 0 '\000'},
            createNotify = {pad00 = 1108060959, parent = 0, window = 2097154, x = 70, y = 0, width = 74, height = 0, borderWidth = 46009, override = 4 '\004', bpad = 0 '\000'}, destroyNotify = {pad00 = 1108060959, event = 0,
              window = 2097154}, unmapNotify = {pad00 = 1108060959, event = 0, window = 2097154, fromConfigure = 70 'F', pad1 = 0 '\000', pad2 = 0 '\000', pad3 = 0 '\000'}, mapNotify = {pad00 = 1108060959, event = 0, window = 2097154,
              override = 70 'F', pad1 = 0 '\000', pad2 = 0 '\000', pad3 = 0 '\000'}, mapRequest = {pad00 = 1108060959, parent = 0, window = 2097154}, reparent = {pad00 = 1108060959, event = 0, window = 2097154, parent = 70, x = 74, y = 0,
              override = 185 '\271', pad1 = 179 '\263', pad2 = 4 '\004', pad3 = 0 '\000'}, configureNotify = {pad00 = 1108060959, event = 0, window = 2097154, aboveSibling = 70, x = 74, y = 0, width = 46009, height = 4,
              borderWidth = 39952, override = 157 '\235', bpad = 254 '\376'}, configureRequest = {pad00 = 1108060959, parent = 0, window = 2097154, sibling = 70, x = 74, y = 0, width = 46009, height = 4, borderWidth = 39952,
              valueMask = 65181, pad1 = 22065}, gravity = {pad00 = 1108060959, event = 0, window = 2097154, x = 70, y = 0, pad1 = 74, pad2 = 308153, pad3 = 4271741968, pad4 = 22065}, resizeRequest = {pad00 = 1108060959, window = 0,
              width = 2, height = 32}, circulate = {pad00 = 1108060959, event = 0, window = 2097154, parent = 70, place = 74 'J', pad1 = 0 '\000', pad2 = 0 '\000', pad3 = 0 '\000'}, property = {pad00 = 1108060959, window = 0,
              atom = 2097154, time = 70, state = 74 'J', pad1 = 0 '\000', pad2 = 0}, selectionClear = {pad00 = 1108060959, time = 0, window = 2097154, atom = 70}, selectionRequest = {pad00 = 1108060959, time = 0, owner = 2097154,
              requestor = 70, selection = 74, target = 308153, property = 4271741968}, selectionNotify = {pad00 = 1108060959, time = 0, requestor = 2097154, selection = 70, target = 74, property = 308153}, colormap = {pad00 = 1108060959,
              window = 0, colormap = 2097154, new = 70 'F', state = 0 '\000', pad1 = 0 '\000', pad2 = 0 '\000'}, mappingNotify = {pad00 = 1108060959, request = 0 '\000', firstKeyCode = 0 '\000', count = 0 '\000', pad1 = 0 '\000'},
            clientMessage = {pad00 = 1108060959, window = 0, u = {l = {type = 2097154, longs0 = 70, longs1 = 74, longs2 = 308153, longs3 = -23225328, longs4 = 22065}, s = {type = 2097154, shorts0 = 70, shorts1 = 0, shorts2 = 74,
                  shorts3 = 0, shorts4 = -19527, shorts5 = 4, shorts6 = -25584, shorts7 = -355, shorts8 = 22065, shorts9 = 0}, b = {type = 2097154, bytes = "F\000\000\000J\000\000\000\271\263\004\000\020\234\235\376\061V\000"}}}}}
#3  0x00005631fe9d8872 in vncHandleClipboardData (data=0x56320139a9c0 "dpkg-query -Wf '${Installed-Size}\\t${Package}\\n' | sort -n") at ../../../../../unix/xserver/hw/vnc/vncSelection.c:174
        rc = <optimized out>
        event = {u = {u = {type = 144 '\220', detail = 195 '\303', sequenceNumber = 143}, keyButtonPointer = {pad00 = 9421712, time = 22066, root = 12582944, event = 0, child = 1326689464, rootX = 32767, rootY = 0, eventX = 20064,
              eventY = 177, state = 0, sameScreen = 0 '\000', pad1 = 0 '\000'}, enterLeave = {pad00 = 9421712, time = 22066, root = 12582944, event = 0, child = 1326689464, rootX = 32767, rootY = 0, eventX = 20064, eventY = 177, state = 0,
              mode = 0 '\000', flags = 0 '\000'}, focus = {pad00 = 9421712, window = 22066, mode = 32 ' ', pad1 = 0 '\000', pad2 = 192 '\300', pad3 = 0 '\000'}, expose = {pad00 = 9421712, window = 22066, x = 32, y = 192, width = 0,
              height = 0, count = 44216, pad2 = 20243}, graphicsExposure = {pad00 = 9421712, drawable = 22066, x = 32, y = 192, width = 0, height = 0, minorEvent = 44216, count = 20243, majorEvent = 255 '\377', pad1 = 127 '\177',
              pad2 = 0 '\000', pad3 = 0 '\000'}, noExposure = {pad00 = 9421712, drawable = 22066, minorEvent = 32, majorEvent = 192 '\300', bpad = 0 '\000'}, visibility = {pad00 = 9421712, window = 22066, state = 32 ' ', pad1 = 0 '\000',
              pad2 = 192 '\300', pad3 = 0 '\000'}, createNotify = {pad00 = 9421712, parent = 22066, window = 12582944, x = 0, y = 0, width = 44216, height = 20243, borderWidth = 32767, override = 0 '\000', bpad = 0 '\000'},
            destroyNotify = {pad00 = 9421712, event = 22066, window = 12582944}, unmapNotify = {pad00 = 9421712, event = 22066, window = 12582944, fromConfigure = 0 '\000', pad1 = 0 '\000', pad2 = 0 '\000', pad3 = 0 '\000'}, mapNotify = {
              pad00 = 9421712, event = 22066, window = 12582944, override = 0 '\000', pad1 = 0 '\000', pad2 = 0 '\000', pad3 = 0 '\000'}, mapRequest = {pad00 = 9421712, parent = 22066, window = 12582944}, reparent = {pad00 = 9421712,
              event = 22066, window = 12582944, parent = 0, x = -21320, y = 20243, override = 255 '\377', pad1 = 127 '\177', pad2 = 0 '\000', pad3 = 0 '\000'}, configureNotify = {pad00 = 9421712, event = 22066, window = 12582944,
              aboveSibling = 0, x = -21320, y = 20243, width = 32767, height = 0, borderWidth = 20064, override = 177 '\261', bpad = 0 '\000'}, configureRequest = {pad00 = 9421712, parent = 22066, window = 12582944, sibling = 0,
              x = -21320, y = 20243, width = 32767, height = 0, borderWidth = 20064, valueMask = 177, pad1 = 0}, gravity = {pad00 = 9421712, event = 22066, window = 12582944, x = 0, y = 0, pad1 = 1326689464, pad2 = 32767, pad3 = 11619936,
              pad4 = 0}, resizeRequest = {pad00 = 9421712, window = 22066, width = 32, height = 192}, circulate = {pad00 = 9421712, event = 22066, window = 12582944, parent = 0, place = 184 '\270', pad1 = 172 '\254', pad2 = 19 '\023',
              pad3 = 79 'O'}, property = {pad00 = 9421712, window = 22066, atom = 12582944, time = 0, state = 184 '\270', pad1 = 172 '\254', pad2 = 20243}, selectionClear = {pad00 = 9421712, time = 22066, window = 12582944, atom = 0},
            selectionRequest = {pad00 = 9421712, time = 22066, owner = 12582944, requestor = 0, selection = 1326689464, target = 32767, property = 11619936}, selectionNotify = {pad00 = 9421712, time = 22066, requestor = 12582944,
              selection = 0, target = 1326689464, property = 32767}, colormap = {pad00 = 9421712, window = 22066, colormap = 12582944, new = 0 '\000', state = 0 '\000', pad1 = 0 '\000', pad2 = 0 '\000'}, mappingNotify = {pad00 = 9421712,
              request = 50 '2', firstKeyCode = 86 'V', count = 0 '\000', pad1 = 0 '\000'}, clientMessage = {pad00 = 9421712, window = 22066, u = {l = {type = 12582944, longs0 = 0, longs1 = 1326689464, longs2 = 32767, longs3 = 11619936,
                  longs4 = 0}, s = {type = 12582944, shorts0 = 0, shorts1 = 0, shorts2 = -21320, shorts3 = 20243, shorts4 = 32767, shorts5 = 0, shorts6 = 20064, shorts7 = 177, shorts8 = 0, shorts9 = 0}, b = {type = 12582944,
                  bytes = "\000\000\000\000\270\254\023O\377\177\000\000`N\261\000\000\000\000"}}}}}
#4  0x00005631fe9da90b in XserverDesktop::requestClipboard (this=<optimized out>) at ../../../../../unix/xserver/hw/vnc/XserverDesktop.cc:184
No locals.
#5  0x00005631fe9cfb66 in vncRequestClipboard () at ../../../../../unix/xserver/hw/vnc/vncExtInit.cc:291
        scr = <optimized out>
#6  0x00005631fe9d79a4 in vncConvertSelection (client=client@entry=0x5632008fc390, selection=70, target=<optimized out>, property=476, requestor=12582944, time=3594488609, data=0x0) at ../../../../../unix/xserver/hw/vnc/vncSelection.c:360
        vdt = <optimized out>
        pSel = 0x563200d56880
        pWin = 0x563201236d50
        rc = <optimized out>
        realProperty = <optimized out>
        event = {u = {u = {type = 144 '\220', detail = 195 '\303', sequenceNumber = 143}, keyButtonPointer = {pad00 = 9421712, time = 22066, root = 12582944, event = 0, child = 1326689600, rootX = 32767, rootY = 0, eventX = 19256,
              eventY = -334, state = 22065, sameScreen = 0 '\000', pad1 = 0 '\000'}, enterLeave = {pad00 = 9421712, time = 22066, root = 12582944, event = 0, child = 1326689600, rootX = 32767, rootY = 0, eventX = 19256, eventY = -334,
              state = 22065, mode = 0 '\000', flags = 0 '\000'}, focus = {pad00 = 9421712, window = 22066, mode = 32 ' ', pad1 = 0 '\000', pad2 = 192 '\300', pad3 = 0 '\000'}, expose = {pad00 = 9421712, window = 22066, x = 32, y = 192,
              width = 0, height = 0, count = 44352, pad2 = 20243}, graphicsExposure = {pad00 = 9421712, drawable = 22066, x = 32, y = 192, width = 0, height = 0, minorEvent = 44352, count = 20243, majorEvent = 255 '\377',
              pad1 = 127 '\177', pad2 = 0 '\000', pad3 = 0 '\000'}, noExposure = {pad00 = 9421712, drawable = 22066, minorEvent = 32, majorEvent = 192 '\300', bpad = 0 '\000'}, visibility = {pad00 = 9421712, window = 22066, state = 32 ' ',
              pad1 = 0 '\000', pad2 = 192 '\300', pad3 = 0 '\000'}, createNotify = {pad00 = 9421712, parent = 22066, window = 12582944, x = 0, y = 0, width = 44352, height = 20243, borderWidth = 32767, override = 0 '\000',
              bpad = 0 '\000'}, destroyNotify = {pad00 = 9421712, event = 22066, window = 12582944}, unmapNotify = {pad00 = 9421712, event = 22066, window = 12582944, fromConfigure = 0 '\000', pad1 = 0 '\000', pad2 = 0 '\000',
              pad3 = 0 '\000'}, mapNotify = {pad00 = 9421712, event = 22066, window = 12582944, override = 0 '\000', pad1 = 0 '\000', pad2 = 0 '\000', pad3 = 0 '\000'}, mapRequest = {pad00 = 9421712, parent = 22066, window = 12582944},
            reparent = {pad00 = 9421712, event = 22066, window = 12582944, parent = 0, x = -21184, y = 20243, override = 255 '\377', pad1 = 127 '\177', pad2 = 0 '\000', pad3 = 0 '\000'}, configureNotify = {pad00 = 9421712, event = 22066,
              window = 12582944, aboveSibling = 0, x = -21184, y = 20243, width = 32767, height = 0, borderWidth = 19256, override = 178 '\262', bpad = 254 '\376'}, configureRequest = {pad00 = 9421712, parent = 22066, window = 12582944,
              sibling = 0, x = -21184, y = 20243, width = 32767, height = 0, borderWidth = 19256, valueMask = 65202, pad1 = 22065}, gravity = {pad00 = 9421712, event = 22066, window = 12582944, x = 0, y = 0, pad1 = 1326689600,
              pad2 = 32767, pad3 = 4273097528, pad4 = 22065}, resizeRequest = {pad00 = 9421712, window = 22066, width = 32, height = 192}, circulate = {pad00 = 9421712, event = 22066, window = 12582944, parent = 0, place = 64 '@',
              pad1 = 173 '\255', pad2 = 19 '\023', pad3 = 79 'O'}, property = {pad00 = 9421712, window = 22066, atom = 12582944, time = 0, state = 64 '@', pad1 = 173 '\255', pad2 = 20243}, selectionClear = {pad00 = 9421712, time = 22066,
              window = 12582944, atom = 0}, selectionRequest = {pad00 = 9421712, time = 22066, owner = 12582944, requestor = 0, selection = 1326689600, target = 32767, property = 4273097528}, selectionNotify = {pad00 = 9421712,
              time = 22066, requestor = 12582944, selection = 0, target = 1326689600, property = 32767}, colormap = {pad00 = 9421712, window = 22066, colormap = 12582944, new = 0 '\000', state = 0 '\000', pad1 = 0 '\000', pad2 = 0 '\000'},
            mappingNotify = {pad00 = 9421712, request = 50 '2', firstKeyCode = 86 'V', count = 0 '\000', pad1 = 0 '\000'}, clientMessage = {pad00 = 9421712, window = 22066, u = {l = {type = 12582944, longs0 = 0, longs1 = 1326689600,
                  longs2 = 32767, longs3 = -21869768, longs4 = 22065}, s = {type = 12582944, shorts0 = 0, shorts1 = 0, shorts2 = -21184, shorts3 = 20243, shorts4 = 32767, shorts5 = 0, shorts6 = 19256, shorts7 = -334, shorts8 = 22065,
                  shorts9 = 0}, b = {type = 12582944, bytes = "\000\000\000\000@\255\023O\377\177\000\000\070K\262\376\061V\000"}}}}}
#7  0x00005631fe9d846a in vncProcConvertSelection (client=0x5632008fc390) at ../../../../../unix/xserver/hw/vnc/vncSelection.c:425
        paramsOkay = <optimized out>
        pWin = 0x563201236d50
        pSel = 0x563200d56880
        rc = 0
        stuff = <optimized out>
#8  0x00005631fea23533 in Dispatch () at ../../../../unix/xserver/dix/dispatch.c:478
        result = <optimized out>
        client = 0x5632008fc390
        start_tick = 27050
#9  0x00005631fea274b4 in dix_main (argc=19, argv=0x7fff4f13af38, envp=<optimized out>) at ../../../../unix/xserver/dix/main.c:276
        i = <optimized out>
        alwaysCheckForInput = {0, 1}
#10 0x00007f23e0796cca in __libc_start_main (main=0x5631fe8fba20 <main>, argc=19, argv=0x7fff4f13af38, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fff4f13af28) at ../csu/libc-start.c:308
        self = <optimized out>
        result = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {0, 889761983230163835, 94772724224112, 0, 0, 0, 6900766666193681275, 6808647058819562363}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x13, 0x7fff4f13af38}, data = {prev = 0x0,
              cleanup = 0x0, canceltype = 19}}}
        not_first_call = <optimized out>
#11 0x00005631fe8fd09a in _start () at /usr/include/c++/9/bits/stl_tree.h:211
```

**Comment by CendioOssman** [comment]
I'm afraid that is an entirely different issue as your trace is in code specific to Xvnc. Please file a new entry here on Github with your details. Also please remove/edit your comment here so we can avoid things that aren't related to the original issue.
